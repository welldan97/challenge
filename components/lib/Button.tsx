import omit from 'lodash/fp/omit';
import Link from 'next/link';
import { ComponentPropsWithoutRef, memo } from 'react';
import { StyledAnchor, StyledButton } from './Button.styles';

// SECTION: Main

interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
  type?: ComponentPropsWithoutRef<'button'>['type'];
}

interface AnchorProps extends ComponentPropsWithoutRef<'a'> {
  type: 'anchor';
  href: string;
}

const Button = memo((props: ButtonProps | AnchorProps = { type: 'button' }) => {
  if (props.type === 'anchor') {
    const passedHref = typeof props.href === 'string' ? props.href : '';
    return (
      <Link href={passedHref}>
        <StyledAnchor {...omit(['type'], props)} />
      </Link>
    );
  }

  return <StyledButton {...props} />;
});

export default Button;
