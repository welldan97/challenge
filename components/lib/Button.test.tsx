import { render } from '@testing-library/react';
import Button from './Button';

// SECTION: Main

describe('Button', () => {
  test('render as button', () => {
    const { container } = render(<Button>test</Button>);

    expect(container).toMatchSnapshot();
  });

  test('render as anchor', () => {
    const { container } = render(
      <Button type="anchor" href="/">
        test
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });
});
