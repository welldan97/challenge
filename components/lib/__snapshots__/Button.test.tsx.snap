// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render as anchor 1`] = `
<div>
  <a
    class="sc-9b79432-1 fFRuBw"
    href="/"
  >
    test
  </a>
</div>
`;

exports[`Button render as button 1`] = `
<div>
  <button
    class="sc-9b79432-0 hZUqsl"
  >
    test
  </button>
</div>
`;
